### PR TITLE
Comments to vue

### PR DESCRIPTION
**Pull request by skjnldsv**
Ref https://github.com/nextcloud/server/issues/20020

- [x] Move to vue
	- [x] New
	- [x] Edit
	- [x] Delete
	- [x] Implement RichCondenteditable component
- [x] Fixed
	- [x] Mentions
- [x] Added features
	- [x] Retry load more comments on failure
	- [x] Undo comments deletion
	- [x] Make exportable and available as full api (see below)
- [x] Requires a new vue release with
	https://github.com/nextcloud/nextcloud-vue/pull/1485

___

## How to use the comments app in your app
1. Load the comments app
	```php
	use OCP\Comments\ICommentsManager;

	/** @var ICommentsManager */
	private $commentsManager;

	$this->commentsManager->load()
	```
2. Create a new instance
	```js
	const commentsView = new OCA.Comments.View('announcement')
	await commentsView.update(ressourceId)
	commentsView.$mount(el)
	```
3. Profit

### Cleanups
I removed all the tab leftovers, the only thing left is the filesPlugin (inline comment count)

### CODE REVIEW

File: apps/comments/src/services/DavClient.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/**
+ * @copyright Copyright (c) 2019 John Molakvoæ <skjnldsv@protonmail.com>
```
Comment (kesselb) [inline apps/comments/src/services/DavClient.js:2]: It seems this file is in your stash for a while ;)

Comment (skjnldsv) [reply apps/comments/src/services/DavClient.js:2]: Shhhhh :zipper_mouth_face:  :see_no_evil: :grin:

File: apps/comments/lib/Event/LoadCommentsApp.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+namespace OCA\Comments\Event;
+
+use OCP\EventDispatcher\Event;
+
+/**
+ * This event is used to load and init the comments app
+ *
+ * @since 21.0.0
+ */
+class LoadCommentsApp extends Event {
```
Comment (ChristophWurst) [inline apps/comments/lib/Event/LoadCommentsApp.php:34]: 1) Not a public API
2) This is a command, not an event

IMO there should be a tiny comments API for this. Like ``\OCP\Comments\ICommentsManager::load`` or ``\OCP\Comments\ICommentsManager::injectIntoPage``

**Review by ChristophWurst — DISMISSED** [review]

File: apps/comments/lib/Event/LoadCommentsApp.php (hunk 2 above)
Comment (skjnldsv) [reply apps/comments/lib/Event/LoadCommentsApp.php:34]: > IMO there should be a tiny comments API for this. Like `\OCP\Comments\ICommentsManager::load` or `\OCP\Comments\ICommentsManager::injectIntoPage`

You're the php bootstrap expert! I was told to do it like that a while back with Viewer or Sidebar, I'll be happy to have a better way :)

Comment (ChristophWurst) [reply apps/comments/lib/Event/LoadCommentsApp.php:34]: It's really not about bootstrap but what to use events for and what not. When you emit the event here you definitely expect something to happen. You're not signaling that something happened to let others react – that would be a true event.

We have the comments manager already. So this is just about adding a new method to the interface and implementing the few lines in the manager.

File: lib/public/Comments/ICommentsManager.php [hunk 3]
```diff
@@ -322,4 +322,11 @@ public function registerDisplayNameResolver($type, \Closure $closure);
 	 * provided ID is unknown. It must be ensured that a string is returned.
 	 */
 	public function resolveDisplayName($type, $id);
+
+	/**
+	 * Load the Comments app into the page
+	 *
+	 * @since 21.0.0
+	 */
+	public function load();
```
Comment (ChristophWurst) [inline lib/public/Comments/ICommentsManager.php:331]:
```suggestion
	public function load(): void;
```

File: apps/comments/lib/AppInfo/Application.php [hunk 4]
```diff
@@ -70,6 +72,10 @@ public function register(IRegistrationContext $context): void {
 			LoadSidebar::class,
 			LoadSidebarScripts::class
 		);
+		$context->registerEventListener(
```
Comment (ChristophWurst) [inline apps/comments/lib/AppInfo/Application.php:75]: :fire:

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

File: apps/comments/lib/AppInfo/Application.php (hunk 4 above)
Comment (skjnldsv) [reply apps/comments/lib/AppInfo/Application.php]: Damn! Thanks

File: package.json [hunk 5]
```diff
@@ -39,12 +39,13 @@
     "@nextcloud/password-confirmation": "^1.0.1",
     "@nextcloud/paths": "^1.1.2",
     "@nextcloud/router": "^1.1.0",
-    "@nextcloud/vue": "^2.7.0",
+    "@nextcloud/vue": "^3.0.0",
```
Comment (ChristophWurst) [inline package.json:42]: sneaky

Comment (rullzer) [reply package.json:42]: :snake:-y

Comment (skjnldsv) [reply package.json:42]: Is it?
Capture d’écran_2020-10-20_14-54-48

**Review by ChristophWurst — APPROVED** [review]

File: package.json (hunk 5 above)
Comment (ChristophWurst) [reply package.json:42]: yes

**Comment by skjnldsv** [comment]
Only took 16 minutes xD
Capture d’écran_2020-10-20_15-08-09

File: package.json (hunk 5 above)
Comment (skjnldsv) [reply package.json:42]: IS IT?
Capture d’écran_2020-10-20_15-09-11

**Comment by skjnldsv** [comment]
Raaah, missing a lib update, final push™

**Comment by skjnldsv** [comment]
Tests needs fixing, but the tests looks awful :cry:

**Review by rullzer — DISMISSED** [review]
Pressing enter doesn't submit the comment

**Comment by skjnldsv** [comment]
> Tests needs fixing, but the tests looks awful

Fixed

> Pressing enter doesn't submit the comment

Fixed

File: lib/private/Collaboration/Collaborators/UserPlugin.php [hunk 6]
```diff
@@ -178,6 +180,8 @@ public function search($search, $limit, $offset, ISearchResult $searchResult) {
 				if ($addToWideResults) {
 					$result['wide'][] = [
 						'label' => $userDisplayName,
+						'subline' => $status['message'],
```
Comment (skjnldsv) [inline lib/private/Collaboration/Collaborators/UserPlugin.php:183]: php 7.3 NODB tests are screaming 
```
1) Test\Collaboration\Collaborators\UserPluginTest::testSearch with data set #4 ('test', false, true, array(), array(), array(array('Test', array(0, 'test'), 'icon-user', null, array())), array(), true, Mock_IUser_27cf199a Object (...))
133 | Undefined index: message
134
```

What is the best way to make sure we have 'message' here @ChristophWurst @rullzer  ?

File: package.json (hunk 5 above)
Comment (ChristophWurst) [reply package.json:42]: for the event of a Vue upgrade in Nextcloud I expect a cake delivered to my home so we can celebrate properly

File: lib/private/Collaboration/Collaborators/UserPlugin.php (hunk 6 above)
Comment (ChristophWurst) [reply lib/private/Collaboration/Collaborators/UserPlugin.php:183]: what to you want to ensure? that it's always passed? hard to enforce with the array parameter.

how about a fallback with ``'subline' => $status['message'] ?? ''``?

**Comment by skjnldsv** [comment]
```
[Tue Oct 20 17:17:00 2020] ArgumentCountError: Too few arguments to function OC\Comments\Manager::__construct(), 3 passed in /drone/src/apps/spreed/lib/Chat/CommentsManager.php on line 46 and exactly 4 expected at /drone/src/lib/private/Comments/Manager.php#73
[Tue Oct 20 17:17:00 2020] 127.0.0.1:46422 [200]: /ocs/v2.php/cloud/users/user0
[Tue Oct 20 17:17:00 2020] ArgumentCountError: Too few arguments to function OC\Comments\Manager::__construct(), 3 passed in /drone/src/apps/spreed/lib/Chat/CommentsManager.php on line 46 and exactly 4 expected at /drone/src/lib/private/Comments/Manager.php#73
```

Tests are failing on spreed, require a fix there

**Review by rullzer — APPROVED** [review]

**Pull request merged**